libs/postgres_ffi/wal_generate/src/lib.rs

```diff
@@ -250,6 +253,84 @@ pub fn generate_simple(client: &mut impl postgres::GenericClient) -> Result<PgLs
     })
 }
 
+pub fn generate_last_wal_record_xlog_switch(
+    client: &mut impl postgres::GenericClient,
+) -> Result<PgLsn> {
+    // Do not use generate_internal because here we end up with flush_lsn exactly on
+    // the segment boundary and insert_lsn after the initial page header, which is unusual.
+    ensure_server_config(client)?;
+
+    client.execute("CREATE table t(x int)", &[])?;
+    let after_xlog_switch: PgLsn = client.query_one("SELECT pg_switch_wal()", &[])?.get(0);
+    let next_segment = PgLsn::from(0x0200_0000);
+    ensure!(
+        after_xlog_switch <= next_segment,
+        "XLOG_SWITCH message ended after the expected segment boundary: {} > {}",
+        after_xlog_switch,
+        next_segment
+    );
+    Ok(next_segment)
+}
+
+pub fn generate_last_wal_record_xlog_switch_ends_on_page_boundary(
+    client: &mut impl postgres::GenericClient,
+) -> Result<PgLsn> {
+    // Do not use generate_internal because here we end up with flush_lsn exactly on
+    // the segment boundary and insert_lsn after the initial page header, which is unusual.
+    ensure_server_config(client)?;
+
+    client.execute("CREATE table t(x int)", &[])?;
+
+    // Add padding so the XLOG_SWITCH record ends exactly on XLOG_BLCKSZ boundary.
+    let base_wal_advance = {
+        let before_lsn = client.pg_current_wal_insert_lsn()?;
+        client.execute(
+            "SELECT pg_logical_emit_message(false, 'swch', REPEAT('a', 10))",
+            &[],
+        )?;
+        (u64::from(client.pg_current_wal_insert_lsn()?) - u64::from(before_lsn)) as usize
+            + XLOG_SIZE_OF_XLOG_RECORD
+    };
+    let mut remaining_lsn =
+        XLOG_BLCKSZ - u64::from(client.pg_current_wal_insert_lsn()?) as usize % XLOG_BLCKSZ;
+    if remaining_lsn < base_wal_advance {
+        remaining_lsn += XLOG_BLCKSZ;
+    }
+    let repeats = 10 + remaining_lsn - base_wal_advance;
```

Comment:
According to this, xlog_switch will *start* on page boundary, not end, isn't it?

Reply:
(see comment above, apparently added `XLOG_SIZE_OF_XLOG_RECORD` adjusts this)

Reply:
That's correct. I will add a comment.